canister/src/unstable_blocks.rs

```diff
@@ -244,43 +276,177 @@ mod test {
         assert_eq!(peek(&forest), Some(&block_0));
         assert_eq!(pop(&mut forest), Some(block_0));
 
-        // Block 1 is now the anchor. It doesn't have children yet,
-        // so calling `pop` should return `None`.
+        // Block 1 is now the anchor. It doesn't have stable
+        // children yet, so calling `pop` should return `None`.
         assert_eq!(peek(&forest), None);
         assert_eq!(pop(&mut forest), None);
     }
 
     #[test]
-    fn forks() {
-        let genesis_block = BlockBuilder::genesis().build();
-        let block = BlockBuilder::with_prev_header(genesis_block.header()).build();
-        let forked_block = BlockBuilder::with_prev_header(genesis_block.header()).build();
+    fn single_chain_various_difficulties() {
+        let block_0 = BlockBuilder::genesis().build().with_mock_difficulty(20);
+        let block_1 = BlockBuilder::with_prev_header(block_0.header())
+            .build()
+            .with_mock_difficulty(15);
+        let block_2 = BlockBuilder::with_prev_header(block_1.header())
+            .build()
+            .with_mock_difficulty(20);
+        let block_3 = BlockBuilder::with_prev_header(block_2.header())
+            .build()
+            .with_mock_difficulty(110);
+
+        let network = Network::Mainnet;
+        let utxos = UtxoSet::new(network);
+        let mut forest = UnstableBlocks::new(&utxos, 7, block_0.clone(), Some(network));
+
+        push(&mut forest, &utxos, block_1.clone()).unwrap();
+        push(&mut forest, &utxos, block_2).unwrap();
+        assert_eq!(peek(&forest), None);
+        assert_eq!(pop(&mut forest), None);
+
+        push(&mut forest, &utxos, block_3).unwrap();
+        // block_0 (the anchor) now has stable child block_1. Because block_1's
+        // difficulty_based_depth is 15 + 20 + 110 = 145 is greater than
+        // normalized_stability_threshold is 20 * 7 = 140 and it does not have
+        // any siblings. Hence, block_0 should be returned when calling `pop`.
```

Comment:
I suggest also documenting these statements with code when possible. For example, the following can help document the fact that block1's d.b. stability threshold is 145:

```
assert_eq!(crate::blocktree::difficulty_based_depth(&forest.tree.children[0], network), 145);
```

(ditto elsewhere)

Reply:
It is done, please give it a look.